packages/ui-components/src/StringArrayInput.tsx

```diff
@@ -0,0 +1,293 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { editIcon } from '@jupyterlab/ui-components';
+import {
+  Button,
+  ButtonGroup,
+  IconButton,
+  InputLabel,
+  List,
+  ListItem,
+  ListItemText,
+  TextField
+} from '@material-ui/core';
+import { FormHelperText } from '@material-ui/core';
+import produce from 'immer';
+import { useCallback, useEffect, useRef, useState } from 'react';
+
+import React from 'react';
+
+import { trashIcon } from './icons';
+
+export interface IArrayInputProps {
+  label: string;
+  defaultValues: string[];
+  description?: string;
+  fieldName?: string;
+  required?: boolean;
+  defaultError?: boolean;
+  placeholder?: string;
+  onChange: (values: string[]) => any;
+}
+
+interface IListItemProps {
+  value?: any;
+  isEditing?: boolean;
+  placeholder?: string;
+  onSubmit?: (value: string) => any;
+  onCancel?: () => any;
+  onDelete?: () => any;
+  onEdit?: () => any;
+}
+
+export const reducer = produce((draft: string[], action) => {
+  const { type, payload } = action;
+  switch (type) {
+    case 'DELETE_ITEM': {
+      const { index } = payload;
+      if (index !== undefined && index < draft.length) {
+        draft.splice(index, 1);
+      }
+      break;
+    }
+    case 'UPSERT_ITEM': {
+      const { index } = payload;
+      if (index !== undefined && index < draft.length) {
+        // If the item is empty remove it.
+        if (payload.value.trim() === '') {
+          draft.splice(index, 1);
+        } else {
+          draft[index] = payload.value;
+        }
+      } else if (payload.value.trim() !== '') {
+        draft.push(payload.value);
+      }
+      break;
+    }
+    case 'UPSERT_ITEMS': {
+      const { index } = payload;
+      if (
+        index !== undefined &&
+        index < draft.length &&
+        payload.items.length > 0
+      ) {
+        // Update value of the selected input with the first value in the array.
+        draft[index] = payload.items[0];
+
+        // Insert the remaining items.
+        draft.splice(index + 1, 0, ...payload.items.slice(1));
+      } else {
+        draft.push(...payload.items);
+      }
+      break;
+    }
+  }
+});
+
+export const ArrayListItem: React.FC<IListItemProps> = ({
+  value,
+  isEditing,
+  placeholder,
+  onSubmit,
+  onCancel,
+  onDelete,
+  onEdit
+}: IListItemProps) => {
+  const inputRef = useRef<HTMLInputElement>(null);
+
+  useEffect(() => {
+    // We want this to be called anytime isEditing becomes true.
+    if (isEditing) {
+      inputRef.current!.focus();
+      inputRef.current!.select();
+    }
+  }, [isEditing]);
+
+  if (isEditing) {
+    return (
+      <div>
+        <TextField
+          inputProps={{ ref: inputRef }}
+          defaultValue={
+            typeof value === 'string'
+              ? value ?? ''
+              : JSON.stringify(value, null, '\t')
```

Comment:
Is there a reason why we are pretty stringifying this? I'm assuming this string will end up in another json string in which case it will end up a 1-liner anyway

Reply:
Oh yeah this is leftover from my settings UI work, I'll remove it cause we don't need it here